### PR TITLE
Allow for source code comments at any level of indentation

### LINKED ISSUE #25043: [JIT] Source code dedent doesn't ignore comments like Python
This can cause issues when you have comments at a lower indentation than the rest of the code:

```
import torch

class Foo(torch.nn.Module):
    def forward(self, x):
        x = torch.neg(x)
# foo foo foo i have a comment at the wrong indent
        return x

torch.jit.script(Foo())
```

```
Traceback (most recent call last):
  File "ought.py", line 9, in <module>
    torch.jit.script(Foo())
  File "/home/jamesreed/pytorch/torch/jit/__init__.py", line 1159, in script
    return torch.jit.torch.jit._recursive.recursive_script(obj)
  File "/home/jamesreed/pytorch/torch/jit/_recursive.py", line 166, in recursive_script
    stubs = list(map(make_stub, filtered_methods))
  File "/home/jamesreed/pytorch/torch/jit/_recursive.py", line 163, in make_stub
    return torch.jit.script_method(func, _jit_internal.createResolutionCallbackFromClosure(func))
  File "/home/jamesreed/pytorch/torch/jit/__init__.py", line 1224, in script_method
    ast = get_jit_def(fn, self_name="ScriptModule")
  File "/home/jamesreed/pytorch/torch/jit/frontend.py", line 164, in get_jit_def
    py_ast = ast.parse(dedent_src)
  File "/home/jamesreed/miniconda3/lib/python3.6/ast.py", line 35, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 1
    def forward(self, x):
    ^
IndentationError: unexpected indent
```

cc @suo

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46548 Allow for source code comments at any level of indentation**

Differential Revision: [D24434778](https://our.internmc.facebook.com/intern/diff/D24434778)

### CODE REVIEW

File: torch/jit/frontend.py [hunk 1]
```diff
@@ -178,34 +179,42 @@ def get_jit_class_def(cls, self_name):
...
+def normalize_source_lines(sourcelines: List[str]) -> List[str]:
     """
-    This is a helper function which checks for multiline strings and
-    indents the strings by calculating the leading space and appending
-    the spaces to each line of the multiline string.The failure to indent
-    multiline strings causes failures during downstream dedent
+    This helper function accepts a list of source lines. It finds the
+    indentation level of the function definition (`def`), then it indents
+    all lines in the function body to a point at or greater than that
+    level. This allows for comments that are at a lower indentation than
```
Comment (jamesr66a) [inline torch/jit/frontend.py:188]: nit: Not just comments but also things like continued string literals

File: torch/jit/frontend.py [hunk 2]
```diff
@@ -178,34 +179,42 @@ def get_jit_class_def(cls, self_name):
...
-            for lines in range(start + 1, end + 1):
-                sourcelines[lines] = ' ' * leading_space + sourcelines[lines]
-    return sourcelines
+
+    def remove_prefix(text, prefix):
+        return text[text.startswith(prefix) and len(prefix):]
+
+    # Find the line and line number containing the function definition
+    for i, l in enumerate(sourcelines):
+        if l.lstrip()[:3] == "def":
```
Comment (jamesr66a) [inline torch/jit/frontend.py:202]: How about `if l.lstrip().startswith('def'):`?

**Review by jamesr66a — APPROVED** [review]
LGTM! Please import and wait for internal tests to pass

**Comment by ppwwyyxx** [comment]
`normalize_source_lines` is not robust: given a function like https://github.com/google/jax/blob/32564442cbab5bec712cadc316a1c01396faf89c/jax/_src/numpy/linalg.py#L296, part of the decorator code also starts with "def", so `normalize_source_lines` will not correctly find the real definition line.

**Pull request closed without merging**